python/shotgun_model/shotgun_query_model.py

```diff
@@ -620,262 +557,271 @@ def _set_tooltip(self, item, data):
         """
-        self._bundle.log_debug("[%s] %s" % (self.__class__.__name__, msg))
+        data_item = self._data_handler.get_data_item_from_uid(uid)
 
-    def _log_warning(self, msg):
-        """
-        Convenience wrapper around warning logging
+        if data_item is None:
+            # no match in data store
+            self._log_debug("...uid '%s' is not part of the data set" % uid)
+            return None
 
-        :param msg: debug message
-        """
-        self._bundle.log_warning("[%s] %s" % (self.__class__.__name__, msg))
+        # this node is loaded in the query cached by the data handler
+        # but may not exist in the model yet - because of deferred loading.
 
-    def _sg_clean_data(self, sg_data):
-        """
-        Recursively clean the supplied SG data for use by clients.
+        # first see if we have it in the model already
+        item = self._get_item_by_unique_id(uid)
+
+        if not item:
+
+            # item was not part of the model. Attempt to load its parents until it is visible.
+            self._log_debug("Item %s does not exist in the tree - will expand tree." % data_item)
 
-        This method currently handles:
+            # now get a list of all parents and recurse back down towards
+            # the node we want to load. If at any level, the data has not
+            # yet been loaded, we expand that level.
+            hierarchy_bottom_up = []
+            node = data_item
+            while node:
+                hierarchy_bottom_up.append(node)
+                node = node.parent
 
-            - Converting datetime objects to universal time stamps.
+            # reverse the list to get the top-down hierarchy
+            hierarchy_top_down = hierarchy_bottom_up[::-1]
 
-        :param sg_data:
-        :return:
+            self._log_debug("Resolved top-down hierarchy to be %s" % hierarchy_top_down)
+
+            for data_item in hierarchy_top_down:
+                # see if we have this item in the tree
+                item = self._get_item_by_unique_id(data_item.unique_id)
+                if not item:
+                    self._log_debug(
+                        "Data item %s does not exist in model - fetching parent's children..." % data_item
+                    )
+                    # this parent does not yet exist in the tree
+                    # find the parent and kick it to expand it
+
+                    # assume that the top level is always loaded in tree
+                    # so that it's always safe to do data_item.parent.uid here
+                    parent_item = self._get_item_by_unique_id(data_item.parent.unique_id)
+                    # get model index
+                    parent_model_index = parent_item.index()
+                    # kick it
+                    self.fetchMore(parent_model_index)
+
+            # now try again
+            item = self._get_item_by_unique_id(uid)
```

Comment:
Is it true that this may return `None` still if the parent items are still being fetched?